### PR TITLE
More speedups for Period comparisons

### PR DESCRIPTION
**Pull request by jbrockmendel**
Following #21582, the biggest avoidable overheads in Period comparisons are in a) `isinstance` calls that can be short-circuited and b) `__ne__` call overhead.  This PR moves `__eq__` and `__ne__` to the cython file, removing the `__ne__` overhead, and changes an `isinstance` check to a try/except.

Using the same profile code from #21582, this brings the `set_index` runtime from 6.603 seconds down to 2.165 seconds.

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -313,6 +313,30 @@ class _BaseOffset(object):
     def __setattr__(self, name, value):
         raise AttributeError("DateOffset objects are immutable.")
 
+    def __eq__(self, other):
+        if is_string_object(other):
+            from pandas.tseries.frequencies import to_offset
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: maybe you want to define a function to return this import, we use this in a number of places?

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -313,6 +313,30 @@ class _BaseOffset(object):
...
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __hash__(self):
+        return hash(self._params)
+
+    @property
+    def _params(self):
+        from pandas.errors import AbstractMethodError
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: can you import this at the top?

**Review by jreback — CHANGES_REQUESTED** [review]
add this issue to your existing perf issue (always do that on follow ons)

File: pandas/_libs/tslibs/offsets.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx]: Wouldn't this just be trading one one-liner for another?

Comment (jreback) [reply pandas/_libs/tslibs/offsets.pyx]: no then then its a function which we can import at the top (and of course if we eventually are able to actually move to_offset to cython its a big win). but its 'cleaner' I think. (certainly can followup on this in new PR). I just don't really like the imports hidden in the functions (as we must do currently), and rather centrailize in a single place.

File: pandas/tseries/offsets.py [hunk 3]
```diff
@@ -182,6 +182,8 @@ def __add__(date):
 
     Since 0 is a bit weird, we suggest avoiding its use.
     """
+    _params = cache_readonly(BaseOffset._params.fget)
+
```
Comment (jreback) [inline pandas/tseries/offsets.py]: can you remove this blank line

File: pandas/_libs/tslibs/offsets.pyx [hunk 4]
```diff
@@ -313,6 +313,37 @@ class _BaseOffset(object):
...
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __hash__(self):
+        return hash(self._params)
+
+    @property  # NB: non-cython subclasses override with cache_readonly
+    def _params(self):
+        all_paras = self.__dict__.copy()
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: can you give an expl of what this is doing (and you can move the comment inside the function too)

**Review by jreback — CHANGES_REQUESTED** [review]
just a minor comment. ping on green.

**Review by jreback — APPROVED** [review]

**Pull request merged**